### ISSUE TITLE
[BuildCheck Suggestion]: Custom BuildCheck reports are not following the canonical error format

### ISSUE THREAD

**Issue by baronfel** [dotnet/msbuild#10786]
### Summary

Custom BuildChecks that report errors don't seem to be emitting with the canonical error format, leading to GitHub Actions not catching the results as expected.

In [this build pipeline](https://github.com/baronfel/sdk-container-demo/actions/runs/11260791412/job/31312798294), @YuliiaKovalova's custom build check detected the following violation and reported it as an error:

```
/usr/share/dotnet/sdk/9.0.100-rc.2.24474.11/Current/Microsoft.Common.props(84,119): error BCC002: https://aka.ms/buildcheck/codes#BCC002 - Detected secret: Username with value: 'runner'
```

In [this pipeline](https://github.com/baronfel/sdk-container-demo/actions/runs/11275795254/job/31358084276) the same check reported as a warning and was caught:

```
/usr/share/dotnet/sdk/9.0.100-rc.2.24474.11/Current/Microsoft.Common.props(84,119): warning BCC002: https://aka.ms/buildcheck/codes#BCC002 - Detected secret: Username with value: 'runner' [/home/runner/work/sdk-container-demo/sdk-container-demo/src/sdk-container-demo/sdk-container-demo.csproj]
```

The gap seems to be that the final project in the square brackets is missing from the error report.

### Background and Motivation

See above

### Sample issue or antipattern that the check should be flagging

See above

**Comment by JanKrivanek** [comment]
This is a weird magic applied by GHA. 

When run purely from MSBuild the diagnostics have identical formatting. On GHA note the extra markup for warnings

   error:
   ```
   2024-10-09T18:21:44.4314294Z /usr/share/dotnet/sdk/9.0.100-rc.2.24474.11/Current/Microsoft.Common.props(84,119): error BCC002: https://aka.ms/buildcheck/codes#BCC002 - Detected secret: Username with value: 'runner'
   ```

   warning:
   ```
   2024-10-10T14:10:45.7254122Z ##[warning]/usr/share/dotnet/sdk/9.0.100-rc.2.24474.11/Current/Microsoft.Common.props(84,119): warning BCC002: https://aka.ms/buildcheck/codes#BCC002 - Detected secret: Username with value: 'runner' [/home/runner/work/sdk-container-demo/sdk-container-demo/src/sdk-container-demo/sdk-container-demo.csproj]
   ```

**Comment by JanKrivanek** [comment]
Possibly - as I'm e.g. not seing those markups for the nuget warnings either.


Though there is one slight differenc between the BuildCheck warnings and errors - warnings are indented while errors are not:

![Image](https://github.com/user-attachments/assets/cf34fa4f-0784-4f81-8216-61ef4150fa3e)


This doesn't apply to 'promoted' warnings to errors. So I suspect warnings get some extra build context that make the console logger indent it as part of that context. I'm digging in...

**Comment by JanKrivanek** [comment]
Ok - I have it.

**tl;dr;:** This is combination of 2 flows:
 * GHA recoginition of build warnings/errors, that requires indentation
 * MSBuild treating of warnings and errors that happens during evaluation - which then cannot be easily promoted/demoted; are not indented under project etc.

**More details:**

The events encountered during evaluation have only EvaluationId, but missing ContextId and InstanceId (more details in doc here: https://github.com/dotnet/msbuild/blob/main/documentation/wiki/Binary-Log.md#meaning-of-various-ids-in-the-buildeventargs). That causes console logger not to indent it under a project execution.

To properly handle `Warnings[Not]As[Errors|Messages]` we bugger BuildCheck warnings encountered during eval and associate execution time BuildContext with them, once that is know - so that LoggingService can properly fetch mappings of warnings promotability/demotability, that is keyed by InstanceId and ContextId. So they can as well be properly indented
